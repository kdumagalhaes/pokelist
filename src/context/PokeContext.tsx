import {
  createContext,
  ReactNode,
  useContext,
  useState,
  useEffect,
  useCallback,
} from 'react'

interface Pokemon {
  id: number
  name: string
  url: string
}

interface PokeProviderProps {
  children: ReactNode
}

interface FavoritePokemon {
  id: number
  name: string
}
interface PokeContextModel {
  apiResponse: Pokemon[]
  getPokePicture: (id: string) => string
  nextPageUrl: string
  prevPageUrl: string
  goToNextPage: () => void
  goToPrevPage: () => void
  getPokeFavorites: (favoritePokemon: FavoritePokemon) => void
  favoritesList: FavoritePokemon[]
}

export const PokeContext = createContext({} as PokeContextModel)

export const PokeProvider = ({ children }: PokeProviderProps) => {
  const [apiResponse, setApiResponse] = useState<Pokemon[]>([])
  const [favoritesList, setFavoritesList] = useState<FavoritePokemon[]>([])
  const [currentPage, setCurrentPage] = useState(
    'https://pokeapi.co/api/v2/pokemon',
  )
  const [nextPageUrl, setNextPageUrl] = useState('')
  const [prevPageUrl, setPrevPageUrl] = useState('')

  // retrieve the basic pokemon data in the general list and pagination url
  useEffect(() => {
    fetch(currentPage)
      .then((res) => res.json())
      .then((json) => {
        setApiResponse(json.results)
        setNextPageUrl(json.next)
        setPrevPageUrl(json.previous)
      })
  }, [currentPage])

  useEffect(() => {
    const favoritesValue = localStorage.getItem('pokefavs')
    if (typeof favoritesValue === 'string') {
      const favorites = JSON.parse(favoritesValue)
      setFavoritesList(favorites)
    }
  }, [])

  const goToNextPage = () => {
    setCurrentPage(nextPageUrl)
  }

  const goToPrevPage = () => {
    setCurrentPage(prevPageUrl)
  }

  // retrieve id and name of the pokemon that the user has clicked on the fav button and save the data in localStore
  const getPokeFavorites = (favoritePokemon: FavoritePokemon) => {
    // if favoritesList doesn't have a pokemon with the same id as favoritePokemon, include it in the list
    if (!favoritesList.find((pokemon) => pokemon.id === favoritePokemon.id)) {
      localStorage.setItem(
        'pokefavs',
        JSON.stringify([...favoritesList, favoritePokemon]),
      )
    }
  }

  // retrieve the pokemon picture based on its id, which is presented in the general pokemon list
  const getPokePicture = useCallback((id: string) => {
    return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/${id}.png`
  }, [])

  const value = {
    apiResponse,
    getPokePicture,
    nextPageUrl,
    prevPageUrl,
    goToNextPage,
    goToPrevPage,
    getPokeFavorites,
    favoritesList,
  }

  return <PokeContext.Provider value={value}>{children}</PokeContext.Provider>
}

export const usePokemon = () => {
  const context = useContext(PokeContext)

  if (context === undefined) {
    throw new Error('usePokemon must be used within PokeContext')
  }

  return context
}
