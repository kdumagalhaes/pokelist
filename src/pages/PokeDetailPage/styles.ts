import styled from 'styled-components'

export const Container = styled.div`
  max-width: 54rem;
  padding: 2rem;
  margin: 3rem auto 15rem auto;
  box-shadow: 0 25px 50px -12px rgb(0 0 0 / 0.25);
  border-radius: 5px;
  display: flex;
  align-items: center;

  .pokemon-pitcure {
    width: 250px;
    height: 250px;
    margin-right: 15px;
  }

  .content {
    width: 100%;
    display: flex;
    flex-direction: column;
    gap: 0.8rem;

    .top-content {
      display: flex;
      align-items: center;
      justify-content: space-between;
      margin-bottom: 0.7rem;

      .pokemon-name {
        font-size: 1.6rem;
        color: ${({ theme }) => theme.darkBlue};
      }

      .fav-btn {
        background: transparent;
        padding: 0.3rem 0.6rem;
        border: none;
        border-radius: 5px;
        color: ${({ theme }) => theme.yellow};
        font-size: 0.8rem;
        text-decoration: none;
        display: flex;
        align-items: center;
        gap: 5px;
        transform: scale(1);
        transition: transform 0.3s;

        &:hover {
          transform: scale(1.1);
        }

        .fav-icon {
          width: 40px;
          height: 40px;
        }
      }
    }

    .pokemon-bio {
      line-height: 24px;
      margin-bottom: 0.5rem;
    }
  }

  .list {
    display: flex;
    gap: 7px;
  }

  .species-value,
  .pokemon-abilities,
  .list,
  .pokemon-name {
    text-transform: capitalize;
  }

  .title {
    font-weight: 700;
    color: ${({ theme }) => theme.darkBlue};
  }

  @media only screen and (max-width: 600px) {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 2rem;
  }
`
