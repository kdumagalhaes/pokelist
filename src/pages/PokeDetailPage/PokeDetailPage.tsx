/* eslint-disable camelcase */
import { Container } from './styles'
import StarIcon from '../../assets/icons/star-pokemon.webp'
import { useLocation } from 'react-router-dom'
import { getPokemonDetails } from '../../utils/pokemonDetails'
import { useEffect, useState } from 'react'

interface PokemonDetailsMode {
  id: number
  name: string
  base_experience: number
  types: [{ type: { name: string } }]
  abilities: [{ ability: { name: string } }]
  height: number
  weight: number
  species: {
    name: string
  }
}

export function PokeDetailPage() {
  const { pathname } = useLocation()
  const [pokemonDetails, setPokemonDetails] = useState({} as PokemonDetailsMode)

  // retrieve data from the api, with the path in the endpoint when the page is load or the pathname variable changes
  useEffect(() => {
    getPokemonDetails(pathname).then((pokeData) => {
      setPokemonDetails(pokeData)
    })
  }, [pathname])

  const {
    id,
    name,
    types,
    abilities,
    height,
    weight,
    species,
    base_experience,
  } = pokemonDetails

  return (
    <Container>
      <img
        src={`https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/${id}.png`}
        alt={name}
        className="pokemon-pitcure"
      />
      <div className="content">
        <div className="top-content">
          <h3 className="pokemon-name">{name}</h3>
          <button className="fav-btn">
            <img src={StarIcon} alt="favorite icon" className="fav-icon" />
          </button>
        </div>
        <div className="pokemon-base-experience">
          <span className="base-experience title">Base experience: </span>
          <span className="base-experience-value">{base_experience}</span>
        </div>
        <ul className="list pokemon-type">
          <span className="type title">Type: </span>
          {types &&
            types.map((data, index) => {
              return <li key={index}>{data.type.name}</li>
            })}
        </ul>
        <ul className="list pokemon-abilities">
          <span className="abilities title">Abilities:</span>
          {abilities &&
            abilities.map((data, index) => {
              return <li key={index}>{data.ability.name}</li>
            })}
        </ul>
        <div className="pokemon-height">
          <span className="height title">Height: </span>
          <span className="height-value">{height}</span>
        </div>
        <div className="pokemon-weight">
          <span className="weight title">Weight: </span>
          <span className="weight-value">{weight}</span>
        </div>
        <div className="pokemon-species">
          <span className="species title">Species: </span>
          <span className="species-value">{species && species.name}</span>
        </div>
      </div>
    </Container>
  )
}
