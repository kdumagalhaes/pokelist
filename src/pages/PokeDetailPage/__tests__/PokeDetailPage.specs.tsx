import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { PokeDetailPage } from '../PokeDetailPage'
import { PokeProvider } from '../../../context/PokeContext'

describe('PokeDetailPage', () => {
  it('initial component should match snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <PokeProvider>
          <PokeDetailPage />
        </PokeProvider>
      </BrowserRouter>,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
