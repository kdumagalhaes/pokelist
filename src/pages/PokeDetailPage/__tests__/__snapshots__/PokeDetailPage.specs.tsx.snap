// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeDetailPage initial component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe fmLMQf"
  >
    <div
      class="sc-gswNZR eDZkQz"
    >
      <img
        class="pokemon-pitcure"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/undefined.png"
      />
      <div
        class="content"
      >
        <div
          class="top-content"
        >
          <h3
            class="pokemon-name"
          />
          <button
            class="fav-btn"
          >
            <img
              alt="favorite icon"
              class="inactive fav-icon"
              src="star-pokemon.webp"
            />
          </button>
        </div>
        <div
          class="pokemon-base-experience"
        >
          <span
            class="base-experience title"
          >
            Base experience: 
          </span>
          <span
            class="base-experience-value"
          />
        </div>
        <ul
          class="list pokemon-type"
        >
          <span
            class="type title"
          >
            Type: 
          </span>
        </ul>
        <ul
          class="list pokemon-abilities"
        >
          <span
            class="abilities title"
          >
            Abilities:
          </span>
        </ul>
        <div
          class="pokemon-height"
        >
          <span
            class="height title"
          >
            Height: 
          </span>
          <span
            class="height-value"
          />
        </div>
        <div
          class="pokemon-weight"
        >
          <span
            class="weight title"
          >
            Weight: 
          </span>
          <span
            class="weight-value"
          />
        </div>
        <div
          class="pokemon-species"
        >
          <span
            class="species title"
          >
            Species: 
          </span>
          <span
            class="species-value"
          />
        </div>
      </div>
    </div>
    <button
      class="back-btn"
    >
      Go back!
    </button>
  </div>
</DocumentFragment>
`;
